edalize/tools/vpr.py

```diff
@@ -90,27 +99,68 @@ def configure(self, edam):
 
         commands = EdaCommands()
 
-        depends = self.name + ".blif"
+        depends = netlist_file
         targets = self.name + ".net"
-        command = ["vpr", arch_xml, self.name + ".blif", "--pack"]
+        command = ["vpr", arch_xml, netlist_file, "--pack"]
         command += sdc_opts + vpr_options
         commands.add(command, [targets], [depends])
 
-        depends = self.name + ".net"
+        # Run generate constraints script of flag set to true
+        if self.tool_options.get("gen_constraints", False):
+            depends = self.name + ".net"
+            targets = "${IOPLACE_FILE}"
+            commands.add(
```

Comment:
I must confess I don't really understand where all these `${SOMETHING}` are defined? I kind of suspect they are supposed to be defined by the f4pga flow but that won't work if vpr is run stand-alone. Let's see for each of these if they can be expanded instead. I assume e.g. that `${OUT_EBLIF}` should be `netlist_file`, or did I misunderstand? Should `${NET_FILE}` be `self.name+.net`? I also think we need to offer the user some way to set `PART`. Should that be a tool option? Or can we read that somehow e.g. directly from the netlist or so?

Reply:
@olofk It's true that this will cause a problem if the gen_constraints feature is run with VPR not being run through the F4PGA flow, but there is no other easy way to have this feature work properly the way F4PGA needs it. 

Some of those variables could be based on the already known stuff like you said, but there are also several variables that are unique to F4PGA and are traditionally set as environmental variables in the Makefile which means they would have to be set in the F4PGA flow, based on the install parameters. 

Previously I had this gen_constraints feature as an array parameter instead of a boolean parameter, where the user is forced to pass in an array of the variables, which prevents them from being hardcoded inside VPR, but it does look and feel a bit more awkward because you have to assume the size of the array. Which option do you think is best?

Reply:
I should add that the PART variable comes from the main F4PGA flow, and is set by the board the user is targeting
 https://github.com/olofk/edalize/blob/7c3d2e01803c7744a89c26de8c0d3836eceb2489/edalize/flows/f4pga.py#L179

Reply:
I understand that they have traditionally been set as environment variables, but is there a good reason for that? I really don't want to make the edalize backend unnecessarily complicated when we have an excellent opportunity to clean up this mess a bit. So please go through them to see which ones are actually needed. I'm confident it will be a much smaller subset than what's currently there.

When you say "based on the install parameters", do you mean that these are absolute paths to files within the the f4pga intallation? If so, that's a questionable design choice and I would highly recommend f4pga to do something about it, but if we have to, we can make it work. There are other backends that require certain environment variables to be set, see .e.g (https://github.com/olofk/edalize/blob/master/edalize/openfpga.py#L67). The important thing here is then to raise an error so that the user know they need to set these variables themselves (like in the openfpga example)

Reply:
> I understand that they have traditionally been set as environment variables, but is there a good reason for that? I really don't want to make the edalize backend unnecessarily complicated when we have an excellent opportunity to clean up this mess a bit. So please go through them to see which ones are actually needed. I'm confident it will be a much smaller subset than what's currently there.

I can go through and get a minimized list of needed variables, but one of the reasons I had so many is because it simplifies the repetition used in each recipe. Do you think regular Make variables would work fine for that or should we just remove variables altogether and just plug the values in everywhere at Makefile generation?

Reply:
> When you say "based on the install parameters", do you mean that these are absolute paths to files within the the f4pga intallation? If so, that's a questionable design choice and I would highly recommend f4pga to do something about it, but if we have to, we can make it work. There are other backends that require certain environment variables to be set, see .e.g (https://github.com/olofk/edalize/blob/master/edalize/openfpga.py#L67). The important thing here is then to raise an error so that the user know they need to set these variables themselves (like in the openfpga example)

There are F4PGA environmental variables, but some of them are set manually by the user at install time. The F4PGA architecture file directory structures are also not consistent across architectures, which I recently noticed and pointed out to the F4PGA people who are working to fix that. I think your suggestion is probably the best way forward, where we assume that the variables shown in the F4PGA install documentation are there, and throw errors if they aren't.